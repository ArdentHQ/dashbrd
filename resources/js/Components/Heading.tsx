import cn from "classnames";
import { createElement, forwardRef, type HTMLAttributes } from "react";
export type HeadingWeight = "bold" | "medium" | "normal";

export type HeadingLevel = 1 | 2 | 3 | 4;

export interface HeadingProperties extends HTMLAttributes<HTMLHeadingElement> {
    level: HeadingLevel;
    weight?: HeadingWeight;
    as?: "h1" | "h2" | "h3" | "h4";
}

const weightClass = (weight: HeadingWeight): string => {
    if (weight === "bold") {
        return "font-bold";
    }

    if (weight === "medium") {
        return "font-medium";
    }

    return "font-normal";
};

const Heading = forwardRef<HTMLHeadingElement, HeadingProperties>(
    ({ level, weight = "medium", as, className, ...properties }, reference): JSX.Element => {
        if (level === 1) {
            return createElement(as ?? "h1", {
                ref: reference,
                className: cn(
                    "break-words text-xl leading-[1.875rem] text-theme-secondary-900 dark:text-theme-dark-50 md:text-2xl md:leading-8 lg:text-[2rem] lg:leading-[2.75rem]",
                    weightClass(weight),
                    className,
                ),
                ...properties,
            });
        }

        if (level === 2) {
            return createElement(as ?? "h2", {
                ref: reference,
                className: cn(
                    "text-xl leading-[1.875rem] text-theme-secondary-900 dark:text-theme-dark-50 md:text-2xl md:leading-8",
                    weightClass(weight),
                    className,
                ),
                ...properties,
            });
        }

        if (level === 3) {
            return createElement(as ?? "h3", {
                ref: reference,
                className: cn(
                    "text-lg leading-7 text-theme-secondary-900 dark:text-theme-dark-50 md:text-xl md:leading-[1.875rem]",
                    weightClass(weight),
                    className,
                ),
                ...properties,
            });
        }

        return createElement(as ?? "h4", {
            ref: reference,
            className: cn(
                "text-base leading-7 text-theme-secondary-900 dark:text-theme-dark-50 sm:text-lg",
                weightClass(weight),
                className,
            ),
            ...properties,
        });
    },
);

Heading.displayName = "Heading";

export { Heading };
