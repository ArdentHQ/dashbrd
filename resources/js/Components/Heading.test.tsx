import React from "react";
import { Heading } from "./Heading";
import { render } from "@/Tests/testing-library";

describe("Heading", () => {
    it("should render h1", () => {
        const { asFragment } = render(<Heading level={1} />);

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render h2", () => {
        const { asFragment } = render(<Heading level={2} />);

        expect(asFragment()).toMatchSnapshot();
    });
    it("should render h3", () => {
        const { asFragment } = render(<Heading level={3} />);

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render h4", () => {
        const { asFragment } = render(<Heading level={4} />);

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render h1 as bold", () => {
        const { asFragment } = render(
            <Heading
                level={1}
                weight="bold"
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render h1 as normal", () => {
        const { asFragment } = render(
            <Heading
                level={1}
                weight="normal"
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render h1 as medium", () => {
        const { asFragment } = render(
            <Heading
                level={1}
                weight="medium"
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
