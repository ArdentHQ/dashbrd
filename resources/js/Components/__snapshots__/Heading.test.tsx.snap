// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Heading > should render h1 1`] = `
<DocumentFragment>
  <h1
    class="break-words text-xl leading-[1.875rem] text-theme-secondary-900 dark:text-theme-dark-50 md:text-2xl md:leading-8 lg:text-[2rem] lg:leading-[2.75rem] font-medium"
  />
</DocumentFragment>
`;

exports[`Heading > should render h1 as bold 1`] = `
<DocumentFragment>
  <h1
    class="break-words text-xl leading-[1.875rem] text-theme-secondary-900 dark:text-theme-dark-50 md:text-2xl md:leading-8 lg:text-[2rem] lg:leading-[2.75rem] font-bold"
  />
</DocumentFragment>
`;

exports[`Heading > should render h1 as medium 1`] = `
<DocumentFragment>
  <h1
    class="break-words text-xl leading-[1.875rem] text-theme-secondary-900 dark:text-theme-dark-50 md:text-2xl md:leading-8 lg:text-[2rem] lg:leading-[2.75rem] font-medium"
  />
</DocumentFragment>
`;

exports[`Heading > should render h1 as normal 1`] = `
<DocumentFragment>
  <h1
    class="break-words text-xl leading-[1.875rem] text-theme-secondary-900 dark:text-theme-dark-50 md:text-2xl md:leading-8 lg:text-[2rem] lg:leading-[2.75rem] font-normal"
  />
</DocumentFragment>
`;

exports[`Heading > should render h2 1`] = `
<DocumentFragment>
  <h2
    class="text-xl leading-[1.875rem] text-theme-secondary-900 dark:text-theme-dark-50 md:text-2xl md:leading-8 font-medium"
  />
</DocumentFragment>
`;

exports[`Heading > should render h3 1`] = `
<DocumentFragment>
  <h3
    class="text-lg leading-7 text-theme-secondary-900 dark:text-theme-dark-50 md:text-xl md:leading-[1.875rem] font-medium"
  />
</DocumentFragment>
`;

exports[`Heading > should render h4 1`] = `
<DocumentFragment>
  <h4
    class="text-base leading-7 text-theme-secondary-900 dark:text-theme-dark-50 sm:text-lg font-medium"
  />
</DocumentFragment>
`;
